### PR TITLE
echotest: add Walk

### PR DESCRIPTION
**Pull request by tbg**
`echotest` is often a good intermediate solution between table-driven
and datadriven testing, when full datadriven testing may be
counter-productive.

However, `echotest.Require` is too centered on tests that make only a
single (or handful) of assertions. For table-driven tests with lots of
cases more automation is helpful. This is now provided by `Walk`.

It would likely be useful to upstream that, but let's try it out here
first.

For more usages, see #89477.

Epic: None

Release note: None

### CODE REVIEW

File: pkg/testutils/echotest/echotest.go [hunk 1]
```diff
@@ -42,3 +51,105 @@ func Require(t *testing.T, act, path string) {
...
+			rm += filepath.Join(w.dir, f) + " "
+		}
+		t.Errorf("some test files are not referenced by a test case, to remove them use:\n%s", rm)
+	}
+	if len(w.missingFiles) > 0 {
+		mk := fmt.Sprintf(`mkdir -p %[1]q && cd %[1]q && echo 'echo'`, w.dir)
+		for _, f := range w.missingFiles {
+			mk += fmt.Sprintf(" \\\n  | tee %q", f)
+		}
+		t.Errorf("some test cases have no test file, to initialize them use the below snippet, then re-run with -rewrite:\n%s", mk)
```
Comment (erikgrinaker) [inline pkg/testutils/echotest/echotest.go:153]: Can't we autogenerate these?

File: pkg/testutils/echotest/echotest_test.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	// subtest and it'll take you to the entry in the slice. This requires
+	// that `t.Run` is invoked directly with the test name (this is a trial
+	// and error finding).
+	for _, test := range []struct{ name string }{
+		{name: "foo"},
+		{name: "bar"},
+	} {
+		t.Run(test.name, w.Do(t, test.name, func(t *testing.T, path string) {
+			Require(t, fmt.Sprintf("hello, %s", test.name), path)
+		}))
```
Comment (erikgrinaker) [inline pkg/testutils/echotest/echotest_test.go:35]: Wouldn't it be better if the callback simply returned the body, leaving the file management to the walker?

**Review by erikgrinaker — COMMENTED** [review]
Maybe I'm missing something, but this seems like a lot of scaffolding for what could be done by a couple of simple helper functions: a test name to file name converter, and a `t.Cleanup` helper for unreferenced files (which admittedly is slightly annoying to make ergonomic).

**Review by herkolategan — CHANGES_REQUESTED** [review]
> Reviewed 6 of 6 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89492)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[`pkg/testutils/echotest/BUILD.bazel` line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89492#-NEG-cvFEqalbLHuO-nR:-NEG-cvFEqalbLHuO-nS:b-2tntj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/18d4450b54f58b2c2246b3e88e2f9e48e44f6146/pkg/testutils/echotest/BUILD.bazel#L16)):*
> ```Starlark
> )
> 
> go_test(
> ```

I think we need to add `data = glob(["testdata/**"])` here for the tests to pass with Bazel.

**Review by tbg — COMMENTED** [review]
> Maybe I'm missing something, but this seems like a lot of scaffolding for what could be done by a couple of simple helper functions: a test name to file name converter, and a t.Cleanup helper for unreferenced files (which admittedly is slightly annoying to make ergonomic).

I went all-in on the ergonomics here, and this isn't the first iteration either, so while we could spend time on this I'm fairly confident it would lead us back to a similar level of scaffolding and will hold off unless you strongly feel that it's worth it. I'm quite happy with how this turned out and would need some insights into what you're trying to improve.
The whole business with tracking and rewriting files isn't anything I want anyone to ever having to think about again.

Leaving fixup commits for now for ease of review. PTAL!

Want to use this for more raft log work soon and while I can stack the PRs, I wouldn't mind getting this in sometime soon.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89492#-:-NGX2FQS0n8pOQN-_DPj:bqsyouy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[`pkg/testutils/echotest/echotest.go` line 153 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89492#-NDhdWNCgV8I4Hwmu4hq-r2-153:-NGWxcyl10db5C44AK1S:bp4vz8w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3be7054a0e867c6ea69e62a12c900cde55ca1398/pkg/testutils/echotest/echotest.go#L153)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Can't we autogenerate these?
</blockquote></details>

You can but you need a trigger. `datadriven` currently doesn't make the `-rewrite` flag accessible from the outside. Suppose I could look into `os.Args` myself. Having to generate the files manually is annoying, even with the copy-paste helper.

^-- went ahead and did the above, worked out quite well. doing this earlier would've saved me some time.
___
*[`pkg/testutils/echotest/echotest_test.go` line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89492#-NDinp2kZ9YjCgNIEzHf-r2-35:-NGWxyx2AYWtvrI2-gCy:b-fznogt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3be7054a0e867c6ea69e62a12c900cde55ca1398/pkg/testutils/echotest/echotest_test.go#L35)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Wouldn't it be better if the callback simply returned the body, leaving the file management to the walker?
</blockquote></details>

Good idea, done.

**Review by erikgrinaker — APPROVED** [review]
>I went all-in on the ergonomics here, and this isn't the first iteration either, so while we could spend time on this I'm fairly confident it would lead us back to a similar level of scaffolding and will hold off unless you strongly feel that it's worth it.

Nah, this is good -- thanks for adding this!

File: pkg/testutils/echotest/echotest_test.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"fmt"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/testutils"
+)
+
+func TestWalk(t *testing.T) {
+	w := Walk(t, testutils.TestDataPath(t))
+	defer w.Check(t)
```
Comment (erikgrinaker) [inline pkg/testutils/echotest/echotest_test.go:23]: nit: do we need `w.Check()` here? Can `Walk()` simply register a `t.Cleanup()` internally?

**Review by herkolategan — APPROVED** [review]
:lgtm:

> Reviewed 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89492#-:-NGrJcLg-eum2OPtjh4f:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

**Review by herkolategan — COMMENTED** [review]
> Reviewed 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89492)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

**Comment by tbg** [comment]
bors r=[erikgrinaker,herkolategan]

**Pull request merged**